RFC: DEVICE_DT_DEFINE_SUB and example of usage with soc-nv-flash device

de-nordic opened the pull request:
The series off commits that:
 1) add DEVICE_DT_DEFINE_SUB that allows to define sub-device which is a device that shares ~~power management and~~ state with parent device (actually it redirects everything to parent device).
 2) define zephyr,null-controller compatible binding that is used in following commits to place soc-nv-flash compatible sub-device under controller in case where soc-nv-flash device is used only for the purpose of compilation.
 3) defines soc-nv-flash generic driver that allows to access flash device by actual node of flash device, not by controller; this makes flash device access uniform with (Q)SPI hanging devices which are addressed directly; the driver is multi-instance driver so it allows several soc-nv-flash devices to co-exist.
4) modifies dts for all devices that have been defining soc-nv-flash node without controller.
5) Moves flash-simulator sample to directly use the soc-nf-flash instead of accessing it via controllre.

de-nordic committed:
drivers/flash/soc-nv-flash: Generic driver for soc-nf-flash nodes

The soc-nv-flash driver that allows to access soc-nv-flash compatible
nodes directly rather than via parent controller device, as it is
done now, for example user may do:
 flash_read(DEVICE_DT_GET(DT_NODELABEL(flash0)), ...)
rather than
 flash_read(DEVICE_DT_GET(DT_PARENT(DT_NODELABEL(flash0))), ...)
making it more aligned with (Q)SPI devices access where device
is referenced directly rather than controller.

The driver is addition which means that both ways of access still
work.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
dts/bindings: Add null-controller to flash controllers

This is special NV-controller that is not a real device and
is used to encapsulate soc-nv-flash devices that are used
at compile time to configure application layout on flash.
Such soc-nv-flash devices are for example used by x86 QEMU.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on drivers/flash/soc_nv_generic.c:
```diff
@@ -0,0 +1,120 @@
+#include <string.h>
+
+/* This driver supports multi-instance and also covers the soc-nv-null device
+ * which is additional compatible to be blaced on soc-nv-flash with no controller,
```
*placed

nordicjm reviewed (commented):
Outside my area but seems OK

de-nordic committed:
device: Provide DEVICE_DT_DEFINE_SUB

Macro allows definition of sub-device which is subsets of functions
of parent device without own state.
Such device will have own API, private data and configuration but
all pm and state operations will be performed on a parent device.

Commit removes static from the definition of Z_DEVICE_STATE_DEFINE
to allow referencing state of parent device.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/drivers/flash_simulator: Switch to using soc-nv-flash device

The commit switches flash_simulator test to directly use soc-nv-flash
device.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
treewide: Add no_flash_controller as parent to soc-nv-flash

The commit adds parent no_flash_controller, compatible to
"Zephyr,null-controller" for devices that have been defining
soc-nv-flash.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

gmarull commented on include/zephyr/device.h:
```diff
@@ -184,6 +184,44 @@ typedef int16_t device_handle_t;
+ * identifier is derived from the node's dependency ordinal. @ref device.name is
+ * set to `DEVICE_DT_NAME(node_id)`.
+ *
+ * Sub-device takes state and power management from parent device, so any
+ * operations affecting power management or state will be applied to parent
+ * device.
```
definitely -1 to state sharing. why can't this be accomplished by sharing some code between separate parent/child PM callbacks?

bjarki-trackunit commented:
Would this not make the flash area offsets opaque to the flash controller? A flash controller can have multiple flash areas (`flash0`, `flash1`, ...) that all have separate offsets, and potentially flash page layouts and area sizes.

How will this work?

de-nordic commented:
> Would this not make the flash area offsets opaque to the flash controller? A flash controller can have multiple flash areas (`flash0`, `flash1`, ...) that all have separate offsets, and potentially flash page layouts and area sizes.
> 
> How will this work?
> 
> A partition within the flash area `flash0` is offset relative to that flash area, a partition in another flash area would have overlapping offsets if the offset of the flash area is not accounted for
> 
> ```
> flash@10000 {
>     boot_partition@0 {
> ...
> 
> flash@20000 {
>     storage_partition@0 {
> ...
> ```
> 
> both partitions have the same offset, 0, but in reality, the real addresses are not the same :)

Partitions are not handled as drivers (so far), but by Flash Map API.
On the other side, if you for example have devices like above you will have to access them through parent:
```
flash_read(DEVICE_DT_GET(DT_PARENT(flash@10000), ...)
flash_read(DEVICE_DT_GET(DT_PARENT(flash@20000), ...)
```
because they are not real devices, controller is. This also means that the controller is actually figuring out where the offset it gets lays. You can already see that the device is the same in both calls, so `offset` in the call has to be positioned within flash node you want to write. This ends up with code like this:
https://github.com/zephyrproject-rtos/zephyr/blob/2fc19aa033c43aab45e11d0c802ebce68d3953f7/drivers/flash/soc_flash_nrf.c#L150-L152, for the controller to figure out, where there could be two independent devices.
If you have sub device then above code would look like this:
```
flash_read(DEVICE_DT_GET(flash@10000, ...)
flash_read(DEVICE_DT_GET(flash@20000, ...)
``
and now you can use offset 0 by both. Additionally for both you can have different compatibilities set.

This does not affect partitions. Partitions are software limitations on devices. In partition map each partition has pointer to driver that serves it, if it is SPI device then it is that device directly, if it is "soc-nv-flash" compatible device, then parent of that device is used.

This means that if you use DT macros instead, instead of Flash Map API, to get device that partition is placed on, you will get different results. For example lets assume that you have `part0` partition placed on SPI `flash0` device:
```
flash0 {
    ...
    partitions {
        part0: part@0 {
        ...       
    }
}
```
to get direct access to the `part0` with Flash API, not Flash Map API, you need to get the device:
```
dev = DEVICE_DT_GET(DT_GPARENT(DT_NODELABEL(part0);
```
that is the device you operate on, the SPI device. If you now make `flash0` label of "device" hanging from "flash-controller", the code above no longer works, because `flash0` is no longer a device. So the same partition definition you have does not work the same, only because you have moved partition definitions. Now, to get device, you need:
```
dev = DEVICE_DT_GET(DT_PARENT(DT_GPARENT(DT_NODELABEL(part0));
```

Note also that most of "soc-nv-flash" devices within Zephyr tree look the same, so you can have basically one generic driver covering these cases.
